man/meteodraw.1
.TH METEOGRAPH "1" "December 2001" "Meteo station tools" Othello
.SH NAME
meteodraw \- draw graphs of meteo data
.SH SYNOPSIS
meteodraw [
.B \-tdVmI
] [
.B \-c dir
] [
.BI \-l logurl
] [
.BI \-e endtime
] [
.BI \-f conffile
] [
.BI \-g graphname
] [
.BI \-p prefix
] [
.BI \-u url
] [
.BI \-L label
] [
.B -x
]
.BI \-s stationname
.IR interval ...
.SH DESCRIPTION
The 
.BR meteodraw (1)
superseeds the
meteograph program present in previous versions of the
.BR meteo (1)
package.
It computes graphs of meteo data for a weather station for various
sampling
.IR interval s.
Allowed values for the
.I interval
are 60, 300, 1800, 7200 and 86400, resulting in graphs spanning a few
hours, a day,
a week, a month and a full year, respectively.
By default, the graph ends at the current point in time, but this
behaviour can be modified by specifying an end timestamp with
the 
.B \-e
option in the format
.IR YYYYMMDDhhmmss .

By default, no graphs are produced, since with the much more flexible configuration
since meteo 0.9.x, there is no resonable default set of graphs.
The
.B \-g
option allows to modify the list of graphs to be produced, the argument to 
.B \-g
is the name of the graph as specified by the name attribute if the graph element.
The graphs are stored in files in the current working directory
unless a different directory is specified with the 
.B \-c
option. The file names are of the
form

.ce
.IR graphname - suffix .png

The
.I graphname
is specified in the configuration file, it is usually, but not necessarily
of the form 
.IR station . parameter ,
examples are 
.I Altendorf.temperature
or
.IR Gurin.humidity .
The suffix is one of
.IR hour ,
.IR day ,
.IR week ,
.I month
or
.IR year ,
for sampling intervals 60, 300, 1800, 7200 and 86400 respectively.
The possible  graph names are summarized in the following  table
If no
.IR interval s
and no
.IR label s
are specified, only the 300, 1800, 7200 and 86400 second interval
graphs are produced. This is because in most cases, 60 second graphs
will require quite different configuration files, as no maximum
and minimum values are available.

.TS
tab(&);
l l.
temperature&outside temperature
temperature_inside&inside temperature
humidity&humidity
humidity_inside&inside humidity
barometer&barometric pressure
wind&wind speed and direction
rain&rain total
radiation&solar and uv radiation
.TE

The parameters defining these graphs are set in the
.BR meteo.xml (5)
file. In the meteograph program, graph dimensions were hardwired,
and intervals of 60 seconds were impossible. Now it is possible
to specify almost any graph format, and even to use dynamically
adapting scales in the configuration file.

The data for the graphs is read from the averages table except for
the 60 second interval graphs, where it is read from the stationdata
table. While it was possible in previous versions to not use the
.BR meteoavg (1)
program and the averages table, this is now required.

.SH OPTIONS
.TP
.BI \-c imagedir
write image files to 
.IR imagedir .
Note that 
.BR meteodraw (1)
performs a
.BR chdir (2)
before opening the configuration file (as specified with the
.B \-f 
option below), so if a relative path for the configuration file
is used, it should be relativ to the image directory.
One possible application of this feature is to have a graph directory
for each weather station, and to put the meteo.xml
file in the same directory as the graphs. 
.BR meteodraw (1) can then be called with

.nf
.ti +3
meteodraw -c graphdir-a -s stationa -f meteo.xml 7200
.ti +3
meteodraw -c graphdir-b -s stationb -f meteo.xml 7200
.fi

to reproduce all two hour graphs for stations
.I a
and
.IR b .
.TP
.B \-d
increase the debug level by one. Debug messages are written to standard
error.
.TP
.BI \-e endtime
compute graphs ending at
.IR endtime .
.TP
.BI \-f conffile
Use 
.I conffile
to read the configuration, see 
.BR meteo.xml (5)
for the configuration syntax.
.TP
.BI \-g graphname
add a graph named
.I graphname
to the list of graphs to be produced.
.TP
.BI \-l logurl
send log messages to the destination described by the 
.IR logurl .
See the
.BR meteo (1)
manual page for the format of log urls.
.TP
.BI \-s stationname
Name of the station for which to compute the graphs.
.TP
.B \-t
include the timestamp in the file name of the generated image files.
.TP
.BI \-L label
derive the data and resolution from the 
.IR label ,
and compute the requested graphs accordingly. See the
.BR meteo (1)
manual page for information about label formats.

.TP
.BI \-p prefix
add the string
.I prefix
to all filenames of graphs produced.

.TP
.BI \-u url
use the url to construct the browser page urls in the image maps generated
by the -m option

.TP
.B \-m
generate an image map and image link for this graph. This is used in the
browser, where its image maps are produced and included in the HTML page
generated by the PHP script. This was previously done in the PHP script,
duplicating a lot of code that was already present in the C code. It also
meant that image dimensions were fixed twice, in the browser PHP script
and in the C code. Since all

.TP
.B \-x
perform locking of
.BR meteodraw (1)
processes so that only a few processes can run at the same time.
For Systems that have System V semaphores, the number of started processes
and the number of processes actually running in parallel can be configured
through
the
.I queued
and
.I running
attributes respectively of the graphs element in the configuration file.
For systems without System V semaphores, the number of waiting processes
is not limited by 
.BR meteodraw (1),
the processes are simply serialized.

.SH OFFSETS
The time values which are the key to the sdata values have a configurable
offset, as described in 
.BR meteo.sql (5).
When
.BR meteodraw (1)
is asked to retrieve data for a graph, it gets the end points of the
interval for which the data is requested.
These points in time are Unix time, but the sample at time t in the
avg table averages the values from (t - offset) to (t - offset + interval).
For the Lajeado station with offset -3h (-10800), a 300 sec average
with the key 106567600 (00:00 GMT on october 9, 2003) is the
average of data measured between

.in +4
106567600 + 10800 and 106567600 + 10800 + 300
.in -4

i.e. between 03:00 and 03:05 GMT on october 9, 2003. This corresponds
to 00:00 to 00:05 local time.

Therefore, to correct the offset, meteopoll needs to do the following:
If an offset is specified with the graph, instead of asking for
data between tbegin and tend, it asks for data between (tbegin + offset)
and (tend + offset). The points thus collected average data
from (tbegin + offset - offset) = tbegin to
(tend + offset - offset + interval) = (tend + interval), i.e. exactly what
we want. Furthermore, the time keys returned will be t - offset, i.e.
the Unix time when the measurement actually occured.

Again, for the Lajeado station, lets consider the 5min graph for
october 9, 2003. The first sample we need averages the data measured
between 03:00 and 03:05 GMT on october 9, 2003. Due to the three hour
offset, meteopoll has to ask for the samples between 00:00 und 23:59 GMT
to accomplish this. But in the graph, it has to use timekey - offset
to label the sample.

.SH BUGS
Dewpoint computations depend on the temperature unit. But since the
dewpoint routine does not know for which station it was called, it just
takes the first temperature unit specification it finds, and uses that.
This means that if you have one station that uses Fahrenheit, and one
that uses the default of Celsius, and you specify only the Fahrenheit
unit, the dewpoint comutation will assume Fahrenheit instead of Celsius.
Therefore, you should always use one temperature unit only per configuration
file. It is recommended to use the computing features of
.BR meteodraw (1)
described in
BR meteo.xml (5)
to convert to other units where necessary.

.SH "SEE ALSO"
.BR meteo (1)
.BR meteo.xml (5),
.BR meteopoll (1),
.BR meteoavg (1),
.BR meteodequeue (1),
.BR meteojob (1)

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).

--
$Id: NEWS,v 1.32 2009/01/10 19:00:23 afm Exp $


src/meteodraw.cc
/*
 * meteodraw.cc -- create meteo drawings
 *
 * (c) 2003 Dr. Andreas Mueller, Beratung und Entwicklung
 * $Id: meteodraw.cc,v 1.21 2004/03/20 01:15:55 afm Exp $
 */
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <string.h>
#include <time.h>
#include <vector>
#include <set>
#include <algorithm>
#include <Configuration.h>
#include <Graphics.h>
#include <mdebug.h>
#include <Timestamp.h>
#include <printver.h>
#include <errno.h>
#include <iostream>
#include <Timelabel.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <Lock.h>
#include <MeteoException.h>


// some type definitions that will be handy
typedef	std::list<meteo::Timelabel>	timelabellist_t;
typedef std::set<std::string>	stringset_t;
typedef	std::vector<int>	intvector_t;

// some global flags, should rather wrap them into an object
static bool	withtimestamps = false;
static bool	imagemap = false;
static bool	images = true;
static bool	imagetag = false;
time_t		end;
std::string	stationname("undefined");
std::string	prefix;
std::string	url;

// format a time_t as a string
static std::string	timestamp(time_t t) {
	char		tsbuffer[32];
	struct tm	*lt = localtime(&t);
	strftime(tsbuffer, sizeof(tsbuffer), "%Y%m%d%H%M%S", lt);
	return std::string(tsbuffer);
}

// test whether an image file is old, compared to the interval. It is old
// if new data could have arrived in the meantime
static bool	imageIsOld(const std::string& filename, int interval) {
	struct stat	sb;
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "checking age of %s",
		filename.c_str());
	// find out whether file exists
	if (stat(filename.c_str(), &sb) < 0) {
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "file %s does not exist yet",
			filename.c_str());
		return true;
	}

	// compute current time
	time_t	now = time(NULL);
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "file age: %d - %d = %d",
		now, sb.st_ctime, now - sb.st_ctime);
	return ((now - sb.st_ctime) > interval);
}

// readintervals goes through the command line to read the intervals for which
// graphs should be produced, and adds them to the intervals set
static void	readintervals(int argc, char *argv[],
			intvector_t& intervals, bool havelabeled) {
	// read all the remaining arguments, they should be valid interval
	// numbers (check this)
	if (optind == argc) {
		// when optind is argc, there are no arguments left. This means
		// that there aren't any intervals specified. Unless we have
		// some labels set, we need to dd all intervals to the intervals
		// set
		if (!havelabeled) {
			// the 60 seconds interval is not pushed by default
			// because it behaves rather differently: There are
			// no maximum/minimum values in this interval, so
			// in most cases a separate configuration file will
			// be needed. We don't want to have to exclude 60 secs
			// interval manually (would be an incompatible change)
			intervals.push_back(300);
			intervals.push_back(1800);
			intervals.push_back(7200);
			intervals.push_back(86400);
		}
	} else {
		// this is the case where we have read the required intervals
		// from the command line
		for (; optind < argc; optind++) {
			switch (int interval = atoi(argv[optind])) {
			case 60:
			case 300:
			case 1800:
			case 7200:
			case 86400:
				intervals.push_back(interval);
				break;
			default:
				mdebug(LOG_CRIT, MDEBUG_LOG, 0,
					"illegal interval %s", argv[optind]);
				exit(EXIT_FAILURE);
				break;
			}
		}
	}
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "read all remaining arguments");
}

// usage displays a usage message on stdout when an error is found or the
// help option is invoked
static void	usage(void) {
	printf(
"usage: meteodraw [ -dtVmIx ] [ -l logurl ] [ -c dir ] [ -G graph ] \n"
"    [ -s station ] [ -p prefix ] [ -g graph ] [ -f config ] [ -u url ] \n"
"    [ -e endtime ] [ -L label ] [ interval ... ]\n"
"	-d		increase debug level\n"
"	-t		include timestamp in generated file name\n"
"	-s stationname	name of the station for which to compute graphs\n"
"	-V		display version and exit\n"
"	-I		create graphs only of the last update is more than\n"
"			the interval length back\n"
"	-l logfile	write debug to log destionation logurl\n"
"	-L label	compute the image for the label\n"
"	-m              write image map info to stdout\n"
"	-c dir		change directory to dir before writing any image files\n"
"	-g graph	include only graph (can be given multiple times)\n"
"	-f config	use configuration file config\n"
"	-p prefix	store images in files with this prefix\n"
"	-u url		use this URL prefix for image maps\n"
"	-e endtime	specify end time of graph\n"
"	-x		lock processes according to graphs attributes\n");
}

// draw a graph based on a timelabel
void	drawlabeled(const std::string& currentgraph, const meteo::Timelabel& ti) {
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "working on graph %s",
		currentgraph.c_str());
	// write the graph to a file
	std::string	outfilename = prefix + currentgraph + "-"
		+ ti.getString() + ".png";
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "outfilename %s", outfilename.c_str());

	// check age of the graph
	bool	imgold = imageIsOld(outfilename, ti.getInterval());
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "image is %sold",
		(imgold) ? "" : "not ");
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "(images || imgold) = %s",
		(images || imgold) ? "true" : "false");

	// make sure we catch any problems that happen during
	// a call graph generation
	try {
		// compute the graph
		meteo::Graphics	graph(ti.getInterval(), ti.getTime(), true,
			currentgraph, (images || imgold));
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "graph %s constructed",
			currentgraph.c_str());

		if (images || imgold) {
			graph.toFile(outfilename);
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "graph written to %s",
				outfilename.c_str());
		}

		// output image map if available
		if (imagemap) {
			std::cout << graph.mapString(url, stationname);
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "image map written");
		}

		// output the image tag if asked to do so
		if (imagetag) {
			std::cout << graph.imageTagString( outfilename);
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "IMG tag written");
		}
	}
	catch (meteo::MeteoException& ex) {
		// report the problem
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "MeteoException caught: %s, %s",
			ex.getReason().c_str(), ex.getAddinfo().c_str());
	}
}

// get the file name suffix based on the interval
std::string	getsuffix(const int interval) {
	switch (interval) {
		case 60:	return std::string("-hour"); break;
		case 300:	return std::string("-day"); break;
		case 1800:	return std::string("-week"); break;
		case 7200:	return std::string("-month"); break;
		case 86400:	return std::string("-year"); break;
	}
	throw meteo::MeteoException("illegal interval found", "");
}

// draw a graph based on the interval
static void	drawinterval(const std::string& currentgraph, int interval) {
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "processing graph %s, interval %d",
		currentgraph.c_str(), interval);
	// construct the file name
	std::string	outfilename = prefix + currentgraph
		+ getsuffix(interval);

	// add a timestamp if requested
	if (withtimestamps) {
		outfilename += "-" + timestamp(end);
	}

	// add the png extension, we only do png, so we don't have to worry
	// about possible other formats
	outfilename += ".png";
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "outfilename %s", outfilename.c_str());

	// find out whether the image exists, and is less than the interval old
	bool	imgold = imageIsOld(outfilename, interval);

	// again, catch all errors during graph generation
	try {
		// compute the graph
		meteo::Graphics	graph(interval, end, false, currentgraph,
			(images || imgold));

		if (images || imgold)
			graph.toFile(outfilename);
	}
	catch (meteo::MeteoException& ex) {
		// report the problem
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "MeteoException caught: %s, %s",
			ex.getReason().c_str(), ex.getAddinfo().c_str());
	}
}

// the main program, it does the following
// - check the command line arguments, and set various variables according
//   the flags found
// - set up logging
// - open the Configuration
// - check which intervals should be drawn
static int	meteodraw(int argc, char *argv[]) {
	int		c;
	std::string	conffilename(METEOCONFFILE);
	std::string	logurl("file:///-");
	stringset_t	requested;
	intvector_t	intervals;
	timelabellist_t	labeled;
	meteo::Timelabel	tl;
	bool		exclusive = false;

	// remember current time
	time(&end);

	// parse the command line
	while (EOF != (c = getopt(argc, argv, "adc:e:f:g:iIl:L:mp:ts:V?u:x")))
		switch (c) {
		case '?':
			usage();
			exit(EXIT_SUCCESS);
			break;
		case 'l':
			logurl = std::string(optarg);
			break;
		case 'c':
			if (chdir(optarg) < 0) {
				mdebug(LOG_CRIT, MDEBUG_LOG, 0,
					"cannot chdir(%s): %s", optarg,
					strerror(errno));
				exit(EXIT_FAILURE);
			}
			break;
		case 'd':
			debug++;
			break;
		case 'e':
			end = meteo::Timestamp(optarg).getTime();
			if (debug)
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"end timestamp: %d", end);
			break;
		case 't':
			withtimestamps = true;
			break;
		case 'f':
			conffilename = std::string(optarg);
			break;
		case 'g':
			requested.insert(std::string(optarg));
			break;
		case 's':
			stationname = std::string(optarg);
			break;
		case 'V':
			fprintver(stdout);
			exit(EXIT_SUCCESS);
			break;
		case 'i':
			imagetag = true;
			break;
		case 'I':
			images = false;
			break;
		case 'm':
			imagemap = true;
			break;
		case 'L':
			tl = meteo::Timelabel(optarg);
			labeled.push_back(tl);
			break;
		case 'p':
			prefix = std::string(optarg);
			break;
		case 'u':
			url = std::string(optarg);
			break;
		case 'x':
			exclusive = true;
			break;
		default:
			mdebug(LOG_ERR, MDEBUG_LOG, 0, "unkown option %c", c);
			break;
		}

	// set up logging
	mdebug_setup("meteodraw", logurl.c_str());
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "set up logging to %s",
		logurl.c_str());

	// warn if we have no station name, some things will not work, or
	// at least will produce strange results
	if ("undefined" == stationname) {
		fprintf(stderr, "%s: station name not set\n", argv[0]);
		mdebug(LOG_WARNING, MDEBUG_LOG, 0, "station name not specified");
	}

	// make sure we have a configuration file, and initialize the config
	// reader
	meteo::Configuration	conf(conffilename);
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "read configuration %s",
		conffilename.c_str());

	// create lock, so the system cannot be flooded by requests
	meteo::Lock	lock(conf);
	if (exclusive)
		// wait until we can fetch the lock
		if (!lock.enter(true)) {
			mdebug(LOG_ERR, MDEBUG_LOG, 0, "lock failed");
			throw meteo::MeteoException("lock failed", "");
		}

	// read the intervals from the command line
	readintervals(argc, argv, intervals, (labeled.size() == 0));

	// do all the labeled graphs
	timelabellist_t::const_iterator	ti;
	for (ti = labeled.begin(); ti != labeled.end(); ti++) {
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "working on label %s",
			ti->getString().c_str());
		// go through the list of requested graph names
		stringset_t::iterator	j;
		for (j = requested.begin(); j != requested.end(); j++) {
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "processing graph %s",
				j->c_str());
			// actually draw the timelabeled graph
			drawlabeled(*j, *ti);

		}
	}

	// if we had any labeled graphs, we should stop here
	if (labeled.size() != 0) {
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "labeled graphs drawn, "
			"suppressing any interval graphing, exiting");
		if (exclusive)
			lock.leave();
		exit(EXIT_SUCCESS);
	}

	// do the same for all intervals
	intvector_t::iterator	i;
	for (i = intervals.begin(); i != intervals.end(); i++) {
		int	interval = *i;
		stringset_t::iterator	j;
		for (j = requested.begin(); j != requested.end(); j++) {
			std::string	currentgraph = *j;
			// actually draw the graph
			drawinterval(currentgraph, interval);
		}
	}

	// release the lock
	if (exclusive)
		lock.leave();

	exit(EXIT_SUCCESS);
}

int	main(int argc, char *argv[]) {
	try {
		meteodraw(argc, argv);
	} catch (meteo::MeteoException& me) {
		fprintf(stderr, "MeteoException in meteodraw: %s/%s\n",
			me.getReason().c_str(), me.getAddinfo().c_str());
		exit(EXIT_FAILURE);
	}
	exit(EXIT_SUCCESS);
}
